Added hardware PWM for Raspi 1

UEckhardt opened the pull request:
I have added support for the hardware PWM pin of the raspberry PI 1. Theoretically all hardware PWM pins of the BCM2835 should be usable.

`int wiringXPwmSetClock(int pin, uint32_t frequency)`
Set the hardware PWM Clock to frequency. Even if the rpi does not support different clock values for  hardware PWM pins, other platforms may support this, so I have added the pin argument, even if this is not used on the rpi. Default for the clock is 100kHz.

`int wiringXPwmSetRange(int pin, uint32_t range)`
Range for PWM counter, defaults to 1024. On the BCM2835 the range can only be set for two different channels, so setting the range for two pins using the same channel to different values will not work and the last call will overwrite the range for the other pin. 

`int wiringXPwmWrite(int pin, uint32_t val)`
Write a value to the PWM counter.

A test program  wiringx-pwm is also added, you can test for example with  `./wiringx-pwm raspberrypi1b1 1 100000 256 128` This should set the pin 1 to a 50% duty cycle, using 100kHz clock.

UEckhardt committed:
Implement hardware PWM for Raspi 1
Support for PWM on BananaPI R1
Support PWM on broadcom 2836 for Raspberry Pi 2 and 3.

CurlyMoo commented:
Nice work.

A few things.
1. Not sure if stdbool is supported widely so just keep 1 and 0 as true and false.
2. Can you follow the wiringX programming style as much as possible.
3. Didn't know the broadcom supported hardware PWM?
4. If so, doesn't this work on the RPi2 and 3 as well?

Also not sure if the `broadcom2835FunctionSelect` fits my vision to keep things as simple as possible.

UEckhardt commented:
stdbool.h is C99 standard and available since 1999. So I don't think this should give any problems. I think the constants true and false are better readable than 1 and 0.

The broadcom and also at least the Allwinner A10 supports Hardware PWM. My implementation should theoretically also work for the BCM2836, but I have no RPi2 to test this. 

I am currently checking if I can also implement PWM on my BananaPi R1 for the Allwinner A10.

The  `broadcom2835FunctionSelect` is a helper function only and should allow to help to extend later on the SPI interfaces.

Currently I would love to have the GPIO, SPI and PWM functionality for wiringX on my RPI and BananaPi R1, since for the BananaPi R1 all other available gpio libraries are really crap.

CurlyMoo commented:
I think you can manage to add the RPi2 as well based on the docs only. I can test if you want.
The same counts for the A10 and the A31s. Both are really similar. I don't have a BPi 1 but do have a BPi M2 so i can test that.

Are you willing to implement all Broadcom and Allwinner PWM? Then we have a nice idea if the functions are generic enough to handle most cases.

UEckhardt commented:
Am 16.05.2016 um 15:13 schrieb CurlyMoo:

> I think you can manage to add the RPi2 as well based on the docs only. I can
> test if you want.

The BCM2835 and BCM2836 seems to be very similar, so I will try to implement
this (or buy a new RPI3 ;-) )

> The same counts for the A10 and the A31s. Both are really similar. I don't have
> a BPi 1 but do have a BPi M2 so i can test that.
> 
> Are you willing to implement all Broadcom and Allwinner PWM? Then we have a
> nice idea if the functions are generic enough to handle most cases.

I try first to get the Banana Pi to work and then add support for the
RPI2 and Bpi M2.

## 

Ulrich Eckhardt                  http://www.uli-eckhardt.de

Ein Blitzableiter auf dem Kirchturm ist das denkbar stärkste
Misstrauensvotum gegen den lieben Gott. (Karl Krauss)

UEckhardt commented:
So, I have added also PWM support for my Banana Pi R1. On the Banana Pi the Pin 7 is the PWM pin. I am not sure if the clock frequency support is really correct for PBI and RPI, I will check this with my oscilloscope next weekend. 

I also modified `CMakeLists.txt` so that a  cmake -DCMAKE_BUILD_TYPE=Debug is supported. This allows to debug core files with symbols.

UEckhardt commented:
The PWM support for the Raspi is now finished. I have tested it with a  raspberrypi 1B2 and a  raspberrypi 3.

The clock setting for the Allwinner SOC needs some work, since the frequency setting does not work as expected.

CurlyMoo commented:
Take your time :+1:

UEckhardt commented:
The frequency settings for the A10 are tested on my BananaPi R1. The PWM support for the A31s is experimental, since I have no BananaPi M2. Could you please test it?

CurlyMoo commented:
Can you type the command i should run to test?

UEckhardt commented:
Am 22.05.2016 um 20:19 schrieb CurlyMoo:

> Can you type the command i should run to test?

First try ./wiringx-pwm bananapi1 7 1000 32 1, this should
set a connected LED to maximum brightness, then
./wiringx-pwm bananapi1 7 1000 32 16 to set the LED to halve
brightness.

Actually I have encountered, that this logic is the other way
round like on the raspi, where a 32 1 would set the connected LED

## to minimum brightness. I will fix this with the next commit.

Ulrich Eckhardt                  http://www.uli-eckhardt.de

Ein Blitzableiter auf dem Kirchturm ist das denkbar stärkste
Misstrauensvotum gegen den lieben Gott. (Karl Krauss)

CurlyMoo commented:
Pin 7 doesn't work:

```
# LD_LIBRARY_PATH=. ./wiringx-pwm bananapim2 7 1000 32 16
ERROR: Pin 7 (PH9) does not support PWM
```

```
# LD_LIBRARY_PATH=. ./wiringx-pwm bananapim2 23 1000 32 1
# LD_LIBRARY_PATH=. ./wiringx-pwm bananapim2 23 1000 32 16
```

This does, but nothing happens.

UEckhardt commented:
Am 23.05.2016 um 20:56 schrieb CurlyMoo:

> Pin 7 doesn't work:

I had a closer look to the A31s spec. The A31s has a lot of differences against
the A10 for PWM. The A31s has four PWM channels, which can be used on six pins.
Each channel has its own clock divider and a different register access.

I think access to this PWM functions are not a big task. The gpio layout needs
to be corrected and support for four PWM channels needs to be added to
allwinnerA31sPwmSetClock,allwinnerA31sPwmWrite and
allwinnerA31sPwmEnableChannel. Since the different channels are accessed via
different registers instead bit shifting a definition for the channels may look
like:

enum pwm_channel_t {
    PWM_CHANNEL0 = 0x00,
    PWM_CHANNEL1 = 0x10,
    PWM_CHANNEL2 = 0x20,
    PWM_CHANNEL3 = 0x30,
    PWM_NOCHANNEL = 0xff
};

But since I have no access to a device with an A31s, I will revert my changes
for the A31s, but letting this version intact, if someone wants to use this as

## a baseline for supporting the A31s.

Ulrich Eckhardt                  http://www.uli-eckhardt.de

Ein Blitzableiter auf dem Kirchturm ist das denkbar stärkste
Misstrauensvotum gegen den lieben Gott. (Karl Krauss)

CurlyMoo commented:
Can you squash all commits into one?

UEckhardt commented:
Squash is done.

CurlyMoo commented:
Can you also commit under your github account instead of an anonymous one?

UEckhardt commented:
The commit was done under my github account. Do I have overlooked something? since I am not an Github expert.

CurlyMoo commented:
The email found here is properly not associated with your github account:
https://api.github.com/repos/UEckhardt/wiringX/commits/1a6d79e99cf87fae0177b3aa8fbae3d39f4aea08

UEckhardt commented:
I have associated this e-mail address as secondary address now. I hope this is OK now.
